File: Ruby/TheDucciSequence.rb
@history = []
@steps = 1
def ducci arr
    @steps += 1
    new_arr = []
    arr.each_with_index do |v, i|
        second_part = i == arr.length - 1 ? arr[0] : arr[i + 1]
        new_arr.push (v - second_part).abs
    end

    print arr.to_s + "\n"

    if @history.include? new_arr or new_arr.inject(&:+) == 0
        print new_arr.to_s + "\n"
        puts @steps.to_s + " steps"
    else
        @history.push new_arr
        ducci new_arr
    end
end

File: Python/TheDucciSequence.py
#https://www.reddit.com/r/dailyprogrammer/comments/8sjcl0/20180620_challenge_364_intermediate_the_ducci/

history = []
def ducci(arr):
	history.append(arr)
	print arr
	new_array = []
	for i, v in enumerate(arr):
		arg = i + 1 if i < len(arr) - 1 else 0
		new_array.append(abs(v - arr[arg]))
	if new_array not in history:
		ducci(new_array)
	
ducci([0, 653, 1854, 4063])
